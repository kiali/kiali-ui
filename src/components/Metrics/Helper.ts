import assign from 'lodash/fp/assign';
import {
  MonitoringDashboard,
  AllLabelsValues,
  SingleLabelValues,
  LabelDisplayName,
  PromLabel,
  Aggregation,
  AllPromLabelsValues,
  TimeSeries
} from '../../types/Metrics';
import { BaseMetricsOptions } from '../../types/MetricsOptions';
import { MetricsSettingsDropdown, MetricsSettings } from '../MetricsOptions/MetricsSettings';
import MetricsDuration from '../MetricsOptions/MetricsDuration';
import { DurationInSeconds } from '../../types/Common';
import { computePrometheusRateParams } from '../../services/Prometheus';

namespace MetricsHelper {
  export const extractLabelValuesOnSeries = (
    series: TimeSeries[],
    aggregations: Aggregation[],
    extracted: AllLabelsValues
  ): void => {
    series.forEach(ts => {
      Object.keys(ts.metric).forEach(k => {
        const agg = aggregations.find(a => a.label === k);
        if (agg) {
          const value = ts.metric[k];
          let values = extracted.get(agg.displayName);
          if (!values) {
            values = {};
            extracted.set(agg.displayName, values);
          }
          values[value] = true;
        }
      });
    });
  };

  export const extractLabelValues = (
    dashboard: MonitoringDashboard,
    previousValues: AllLabelsValues
  ): AllLabelsValues => {
    // Find all labels on all series
    const labelsWithValues: AllLabelsValues = new Map();
    dashboard.aggregations.forEach(agg => labelsWithValues.set(agg.displayName, {}));
    dashboard.charts.forEach(chart => {
      if (chart.metric) {
        extractLabelValuesOnSeries(chart.metric.matrix, dashboard.aggregations, labelsWithValues);
      }
      if (chart.histogram) {
        Object.keys(chart.histogram).forEach(stat => {
          extractLabelValuesOnSeries(chart.histogram![stat].matrix, dashboard.aggregations, labelsWithValues);
        });
      }
    });
    // Keep existing show flag
    labelsWithValues.forEach((values: SingleLabelValues, key: LabelDisplayName) => {
      const previous = previousValues.get(key);
      if (previous) {
        Object.keys(values).forEach(k => {
          if (previous.hasOwnProperty(k)) {
            values[k] = previous[k];
          }
        });
      }
    });
    return labelsWithValues;
  };

  export const mergeLabelFilter = (
    labelValues: AllLabelsValues,
    label: LabelDisplayName,
    value: string,
    checked: boolean
  ): AllLabelsValues => {
    const newLabels = new Map();
    labelValues.forEach((val, key) => {
      const newVal = assign(val)({});
      if (key === label) {
        newVal[value] = checked;
      }
      newLabels.set(key, newVal);
    });
    return newLabels;
  };

  export const convertAsPromLabels = (aggregations: Aggregation[], labels: AllLabelsValues): AllPromLabelsValues => {
    const promLabels = new Map<PromLabel, SingleLabelValues>();
    labels.forEach((val, k) => {
      const chartLabel = aggregations.find(l => l.displayName === k);
      if (chartLabel) {
        promLabels.set(chartLabel.label, val);
      }
    });
    return promLabels;
  };

  export const settingsToOptions = (
    settings: MetricsSettings,
    opts: BaseMetricsOptions,
    aggregations?: Aggregation[]
  ) => {
    opts.avg = settings.showAverage;
    opts.quantiles = settings.showQuantiles;
    opts.byLabels = [];
    if (aggregations) {
      settings.activeLabels.forEach(lbl => {
        const agg = aggregations.find(a => a.displayName === lbl);
        if (agg) {
          opts.byLabels!.push(agg.label);
        }
      });
    }
  };

  export const initMetricsSettings = (opts: BaseMetricsOptions, aggregations?: Aggregation[]) => {
    settingsToOptions(MetricsSettingsDropdown.initialMetricsSettings(), opts, aggregations);
  };

  export const durationToOptions = (duration: DurationInSeconds, opts: BaseMetricsOptions) => {
    opts.duration = duration;
    const intervalOpts = computePrometheusRateParams(duration);
    opts.step = intervalOpts.step;
    opts.rateInterval = intervalOpts.rateInterval;
  };

  export const initDuration = (opts: BaseMetricsOptions): BaseMetricsOptions => {
    durationToOptions(MetricsDuration.initialDuration(), opts);
    return opts;
  };
}

export default MetricsHelper;
