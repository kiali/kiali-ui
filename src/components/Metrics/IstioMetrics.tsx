import * as React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps, withRouter } from 'react-router';
import { Card, CardBody, Grid, GridItem, Toolbar, ToolbarGroup, ToolbarItem } from '@patternfly/react-core';
import { Dashboard, DashboardModel, ExternalLink, Overlay, VCDataPoint } from '@kiali/k-charted-pf4';
import { style } from 'typestyle';

import RefreshContainer from '../../components/Refresh/Refresh';
import { RenderComponentScroll } from '../../components/Nav/Page';
import * as API from '../../services/Api';
import { KialiAppState } from '../../store/Store';
import { TimeRange, evalTimeRange } from '../../types/Common';
import { Direction, IstioMetricsOptions, Reporter } from '../../types/MetricsOptions';
import * as AlertUtils from '../../utils/AlertUtils';

import * as MetricsHelper from './Helper';
import { MetricsSettings, LabelsSettings } from '../MetricsOptions/MetricsSettings';
import { MetricsSettingsDropdown } from '../MetricsOptions/MetricsSettingsDropdown';
import MetricsReporter from '../MetricsOptions/MetricsReporter';
import history, { URLParam } from '../../app/History';
import { MetricsObjectTypes } from '../../types/Metrics';
import { GrafanaInfo } from '../../types/GrafanaInfo';
import { MessageType } from '../../types/MessageCenter';
import { GrafanaLinks } from './GrafanaLinks';
import { SpanOverlay } from './SpanOverlay';
import TimeRangeComponent from 'components/Time/TimeRangeComponent';
import { retrieveTimeRange, storeBounds } from 'components/Time/TimeRangeHelper';
import { RightActionBar } from 'components/RightActionBar/RightActionBar';

type MetricsState = {
  dashboard?: DashboardModel;
  labelsSettings: LabelsSettings;
  grafanaLinks: ExternalLink[];
  spanOverlay?: Overlay;
  timeRange: TimeRange;
};

type ObjectId = {
  namespace: string;
  object: string;
};

type IstioMetricsProps = ObjectId &
  RouteComponentProps<{}> & {
    objectType: MetricsObjectTypes;
    direction: Direction;
  };

type Props = IstioMetricsProps & {
  // Redux props
  jaegerIntegration: boolean;
};

const displayFlex = style({
  display: 'flex'
});

class IstioMetrics extends React.Component<Props, MetricsState> {
  options: IstioMetricsOptions;
  spanOverlay: SpanOverlay;
  static grafanaInfoPromise: Promise<GrafanaInfo | undefined> | undefined;

  constructor(props: Props) {
    super(props);

    const settings = MetricsHelper.retrieveMetricsSettings();
    const timeRange = retrieveTimeRange() || MetricsHelper.defaultMetricsDuration;
    this.options = this.initOptions(settings);
    // Initialize active filters from URL
    this.state = { labelsSettings: settings.labelsSettings, grafanaLinks: [], timeRange: timeRange };
    this.spanOverlay = new SpanOverlay(changed => this.setState({ spanOverlay: changed }));
  }

  initOptions(settings: MetricsSettings): IstioMetricsOptions {
    const options: IstioMetricsOptions = {
      reporter: MetricsReporter.initialReporter(this.props.direction),
      direction: this.props.direction
    };
    MetricsHelper.settingsToOptions(settings, options);
    return options;
  }

  componentDidMount() {
    this.fetchGrafanaInfo();
    this.refresh();
  }

  refresh = () => {
    this.fetchMetrics();
    if (this.props.jaegerIntegration) {
      this.spanOverlay.fetch(this.props.namespace, this.props.object, this.state.timeRange);
    }
  };

  fetchMetrics = () => {
    // Time range needs to be reevaluated everytime fetching
    MetricsHelper.timeRangeToOptions(this.state.timeRange, this.options);
    let promise: Promise<API.Response<DashboardModel>>;
    switch (this.props.objectType) {
      case MetricsObjectTypes.WORKLOAD:
        promise = API.getWorkloadDashboard(this.props.namespace, this.props.object, this.options);
        break;
      case MetricsObjectTypes.APP:
        promise = API.getAppDashboard(this.props.namespace, this.props.object, this.options);
        break;
      case MetricsObjectTypes.SERVICE:
      default:
        promise = API.getServiceDashboard(this.props.namespace, this.props.object, this.options);
        break;
    }
    return promise
      .then(response => {
        const labelsSettings = MetricsHelper.extractLabelsSettings(response.data, this.state.labelsSettings);
        this.setState({
          dashboard: response.data,
          labelsSettings: labelsSettings
        });
      })
      .catch(error => {
        AlertUtils.addError('Could not fetch metrics.', error);
        throw error;
      });
  };

  fetchGrafanaInfo() {
    if (!IstioMetrics.grafanaInfoPromise) {
      IstioMetrics.grafanaInfoPromise = API.getGrafanaInfo().then(response => {
        if (response.status === 204) {
          return undefined;
        }
        return response.data;
      });
    }
    IstioMetrics.grafanaInfoPromise
      .then(grafanaInfo => {
        if (grafanaInfo) {
          this.setState({ grafanaLinks: grafanaInfo.externalLinks });
        } else {
          this.setState({ grafanaLinks: [] });
        }
      })
      .catch(err => {
        AlertUtils.addError(
          'Could not fetch Grafana info. Turning off links to Grafana.',
          err,
          'default',
          MessageType.INFO
        );
      });
  }

  onMetricsSettingsChanged = (settings: MetricsSettings) => {
    MetricsHelper.settingsToOptions(settings, this.options);
    this.fetchMetrics();
  };

  onLabelsFiltersChanged = (labelsFilters: LabelsSettings) => {
    this.setState({ labelsSettings: labelsFilters });
  };

  onTimeFrameChanged = (range: TimeRange) => {
    this.setState({ timeRange: range }, () => {
      this.spanOverlay.resetLastFetchTime();
      this.refresh();
    });
  };

  onReporterChanged = (reporter: Reporter) => {
    this.options.reporter = reporter;
    this.fetchMetrics();
  };

  onClickDataPoint = (_, datum: VCDataPoint) => {
    if ('traceId' in datum) {
      history.push(
        `/namespaces/${this.props.namespace}/services/${this.props.object}?tab=traces&${URLParam.JAEGER_TRACE_ID}=${datum.traceId}`
      );
    }
  };

  private onDomainChange(dates: [Date, Date]) {
    if (dates && dates[0] && dates[1]) {
      const range: TimeRange = {
        from: dates[0].getTime(),
        to: dates[1].getTime()
      };
      storeBounds(range);
      this.onTimeFrameChanged(range);
    }
  }

  render() {
    const urlParams = new URLSearchParams(history.location.search);
    const expandedChart = urlParams.get('expand') || undefined;

    return (
      <>
        <RightActionBar>
          <TimeRangeComponent
            range={this.state.timeRange}
            onChanged={this.onTimeFrameChanged}
            tooltip={'Time range'}
            allowCustom={true}
          />
          <RefreshContainer id="metrics-refresh" handleRefresh={this.refresh} hideLabel={true} />
        </RightActionBar>
        <RenderComponentScroll>
          <Grid style={{ padding: '10px' }}>
            <GridItem span={12}>
              <Card>
                <CardBody>
                  {this.renderOptionsBar()}
                  {this.state.dashboard && (
                    <Dashboard
                      dashboard={this.state.dashboard}
                      labelValues={MetricsHelper.convertAsPromLabels(this.state.labelsSettings)}
                      expandedChart={expandedChart}
                      expandHandler={this.expandHandler}
                      onClick={this.onClickDataPoint}
                      labelPrettifier={MetricsHelper.prettyLabelValues}
                      overlay={this.state.spanOverlay}
                      timeWindow={evalTimeRange(retrieveTimeRange() || MetricsHelper.defaultMetricsDuration)}
                      brushHandlers={{ onDomainChangeEnd: (_, props) => this.onDomainChange(props.currentDomain.x) }}
                    />
                  )}
                </CardBody>
              </Card>
            </GridItem>
          </Grid>
        </RenderComponentScroll>
      </>
    );
  }

  renderOptionsBar() {
    return (
      <Toolbar style={{ paddingBottom: 8 }}>
        <ToolbarGroup>
          <ToolbarItem>
            <MetricsSettingsDropdown
              onChanged={this.onMetricsSettingsChanged}
              onLabelsFiltersChanged={this.onLabelsFiltersChanged}
              labelsSettings={this.state.labelsSettings}
              hasHistograms={true}
            />
          </ToolbarItem>
        </ToolbarGroup>
        <ToolbarGroup>
          <ToolbarItem className={displayFlex}>
            <MetricsReporter onChanged={this.onReporterChanged} direction={this.props.direction} />
          </ToolbarItem>
        </ToolbarGroup>
        <ToolbarGroup>
          <GrafanaLinks
            links={this.state.grafanaLinks}
            namespace={this.props.namespace}
            object={this.props.object}
            objectType={this.props.objectType}
          />
        </ToolbarGroup>
      </Toolbar>
    );
  }

  private expandHandler = (expandedChart?: string) => {
    const urlParams = new URLSearchParams(history.location.search);
    urlParams.delete('expand');
    if (expandedChart) {
      urlParams.set('expand', expandedChart);
    }
    history.push(history.location.pathname + '?' + urlParams.toString());
  };
}

const mapStateToProps = (state: KialiAppState) => {
  return {
    jaegerIntegration: state.jaegerState ? state.jaegerState.integration : false
  };
};

const IstioMetricsContainer = withRouter<RouteComponentProps<{}> & IstioMetricsProps, any>(
  connect(mapStateToProps)(IstioMetrics)
);

export default IstioMetricsContainer;
