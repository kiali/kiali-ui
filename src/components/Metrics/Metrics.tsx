import * as React from 'react';
import { Alert, Icon } from 'patternfly-react';
import GrafanaInfo from '../../types/GrafanaInfo';
import { authentication } from '../../utils/Authentication';
import * as API from '../../services/Api';
import { computePrometheusQueryInterval } from '../../services/Prometheus';
import MetricsOptionsBar from '../../components/MetricsOptions/MetricsOptionsBar';
import MetricsOptions from '../../types/MetricsOptions';
import HistogramChart from './HistogramChart';
import MetricChart from './MetricChart';
import { Histogram, MetricGroup } from '../../types/Metrics';
import history from '../../app/History';
import { HistoryManager } from '../../app/History';
import { Link } from 'react-router-dom';
import { style } from 'typestyle';

const expandedChartContainerStyle = style({
  height: 'calc(100vh - 248px)'
});

const expandedChartBackLinkStyle = style({
  marginTop: '-1.7em',
  textAlign: 'right'
});

type ChartDefinition = {
  familyName: string;
  component: any;
  metrics?: MetricGroup | Histogram;
};

type MetricsState = {
  alertDetails?: string;
  grafanaLink?: string;
  pollMetrics?: number;
  metricReporter: string;
  metricData?: any;
};

type ObjectId = {
  namespace: string;
  object: string;
};

type MetricsProps = ObjectId & {
  isPageVisible?: boolean;
  objectType: string;
  metricsType: string;
};

class Metrics extends React.Component<MetricsProps, MetricsState> {
  static defaultProps = {
    isPageVisible: true
  };

  options: MetricsOptions;

  constructor(props: MetricsProps) {
    super(props);

    let metricReporter = 'source';

    const metricReporterParam = HistoryManager.getParam('reporter');
    if (metricReporterParam != null) {
      metricReporter = metricReporterParam;
    } else if (this.props.metricsType === 'inbound') {
      metricReporter = 'destination';
    }

    this.state = {
      metricReporter: metricReporter
    };
  }

  getChartsDef(): { [key: string]: ChartDefinition } {
    let inboundCharts = {
      request_count_in: { familyName: 'Request volume (ops)', component: MetricChart },
      request_duration_in: { familyName: 'Request duration (seconds)', component: HistogramChart },
      request_size_in: { familyName: 'Request size (bytes)', component: HistogramChart },
      response_size_in: { familyName: 'Response size (bytes)', component: HistogramChart },
      tcp_received_in: { familyName: 'TCP received (bps)', component: MetricChart },
      tcp_sent_in: { familyName: 'TCP sent (bps)', component: MetricChart }
    };
    let charts: { [key: string]: ChartDefinition } = inboundCharts;

    if (this.props.metricsType === 'outbound') {
      charts = {
        request_count_out: { familyName: 'Request volume (ops)', component: MetricChart },
        request_duration_out: { familyName: 'Request duration (seconds)', component: HistogramChart },
        request_size_out: { familyName: 'Request size (bytes)', component: HistogramChart },
        response_size_out: { familyName: 'Response size (bytes)', component: HistogramChart },
        tcp_received_out: { familyName: 'TCP received (bps)', component: MetricChart },
        tcp_sent_out: { familyName: 'TCP sent (bps)', component: MetricChart }
      };
    }

    if (this.state.metricData) {
      Object.keys(charts).forEach(k => {
        const chart = charts[k];
        const reporter = this.state.metricReporter === 'destination' ? 'dest' : 'source';
        const histo = this.state.metricData[reporter].histograms[k];

        if (histo) {
          chart.metrics = histo;
        } else {
          chart.metrics = this.state.metricData[reporter].metrics[k];
        }
      });
    }

    return charts;
  }

  componentDidMount() {
    this.getGrafanaInfo();
  }

  componentWillUnmount() {
    if (this.state.pollMetrics) {
      clearInterval(this.state.pollMetrics);
    }
  }

  onOptionsChanged = (options: MetricsOptions) => {
    this.options = options;
    options.filters = [
      'request_count',
      'request_size',
      'request_duration',
      'response_size',
      'tcp_received',
      'tcp_sent'
    ];
    const intervalOpts = computePrometheusQueryInterval(options.duration!);
    options.step = intervalOpts.step;
    options.rateInterval = intervalOpts.rateInterval;
    this.fetchMetrics();
    HistoryManager.setParam('duration', options.duration!.toString(10));
  };

  onPollIntervalChanged = (pollInterval: number) => {
    let newRefInterval: number | undefined = undefined;
    if (this.state.pollMetrics) {
      clearInterval(this.state.pollMetrics);
    }
    if (pollInterval > 0) {
      newRefInterval = window.setInterval(this.fetchMetrics, pollInterval);
    }
    this.setState({ pollMetrics: newRefInterval });
    HistoryManager.setParam('pi', pollInterval.toString(10));
  };

  onReporterChanged = (reporter: string) => {
    this.setState({ metricReporter: reporter });
    HistoryManager.setParam('reporter', reporter);
  };

  fetchMetrics = () => {
    let promise;
    switch (this.props.objectType) {
      case 'service':
        promise = API.getServiceMetrics(authentication(), this.props.namespace, this.props.object, this.options);
        break;
      case 'workload':
        promise = API.getWorkloadMetrics(authentication(), this.props.namespace, this.props.object, this.options);
        break;
      case 'app':
        promise = API.getAppMetrics(authentication(), this.props.namespace, this.props.object, this.options);
        break;
      default:
        promise = API.getServiceMetrics(authentication(), this.props.namespace, this.props.object, this.options);
        break;
    }
    Promise.all([promise])
      .then(([response]) => {
        this.setState({ metricData: response.data });
      })
      .catch(error => {
        this.setState({ alertDetails: API.getErrorMsg('Cannot fetch metrics.', error) });
        console.error(error);
      });
  };

  getGrafanaLink(info: GrafanaInfo): string {
    let grafanaLink;
    switch (this.props.objectType) {
      case 'service':
        grafanaLink = `${info.url}${info.serviceDashboardPath}?${info.varService}=${this.props.object}.${
          this.props.namespace
        }.svc.cluster.local`;
        break;
      case 'workload':
        grafanaLink = `${info.url}${info.workloadDashboardPath}?${info.varNamespace}=${this.props.namespace}&${
          info.varWorkload
        }=${this.props.object}`;
        break;
      default:
        grafanaLink = `${info.url}${info.workloadDashboardPath}?${info.varNamespace}=${this.props.namespace}`;
    }
    return grafanaLink;
  }

  getGrafanaInfo = () => {
    API.getGrafanaInfo(authentication())
      .then(response => {
        const info: GrafanaInfo = response['data'];
        if (info) {
          this.setState({
            grafanaLink: this.getGrafanaLink(info)
          });
        } else {
          this.setState({ grafanaLink: undefined });
        }
      })
      .catch(error => {
        this.setState({ grafanaLink: undefined });
        console.error(error);
      });
  };

  dismissAlert = () => this.setState({ alertDetails: undefined });

  render() {
    const urlParams = new URLSearchParams(history.location.search);
    const expandedChart = urlParams.get('expand');
    urlParams.delete('expand');
    const notExpandedLink = history.location.pathname + '?' + urlParams.toString();

    return (
      <div>
        {expandedChart && (
          <h3 className={expandedChartBackLinkStyle}>
            <Link to={notExpandedLink}>
              <Icon name="angle-double-left" /> View all metrics
            </Link>
          </h3>
        )}
        {this.state.alertDetails && <Alert onDismiss={this.dismissAlert}>{this.state.alertDetails}</Alert>}
        <MetricsOptionsBar
          onOptionsChanged={this.onOptionsChanged}
          onPollIntervalChanged={this.onPollIntervalChanged}
          onReporterChanged={this.onReporterChanged}
          onManualRefresh={this.fetchMetrics}
          metricReporter={this.state.metricReporter}
        />
        {expandedChart ? this.renderExpandedChart(expandedChart) : this.renderMetrics()}
      </div>
    );
  }

  renderMetrics() {
    if (!this.props.isPageVisible) {
      return null;
    }

    let charts = this.getChartsDef();

    return (
      <div className="card-pf">
        <div className="row row-cards-pf">
          <div className="col-xs-12">
            <div className="card-pf-accented card-pf-aggregate-status">
              <div className="card-pf-body">
                {Object.keys(charts).map(key => this.renderNormalChart(key, charts[key]))}
              </div>
            </div>
          </div>
          {this.state.grafanaLink && (
            <span id="grafana-link">
              <a href={this.state.grafanaLink} target="_blank">
                View in Grafana
              </a>
            </span>
          )}
        </div>
      </div>
    );
  }

  private renderNormalChart(chartKey: string, chart: ChartDefinition) {
    return this.renderChart(chartKey, chart);
  }

  private renderExpandedChart(chartKey: string) {
    return (
      <div className={expandedChartContainerStyle}>
        {this.renderChart(chartKey, this.getChartsDef()[chartKey], true)}
      </div>
    );
  }

  private renderChart(chartKey: string, chart: ChartDefinition, isExpanded: boolean = false) {
    if (!chart || !chart.metrics) {
      return undefined;
    }
    const props: any = {
      key: chartKey + this.state.metricReporter,
      familyName: chart.familyName
    };
    if ((chart.metrics as MetricGroup).matrix) {
      props.series = (chart.metrics as MetricGroup).matrix;
    } else {
      props.histogram = chart.metrics;
    }

    if (!isExpanded) {
      props.onExpandRequested = () => this.onExpandHandler(chartKey);
    }

    return React.createElement(chart.component, props);
  }

  private onExpandHandler = (chartKey: string): void => {
    const urlParams = new URLSearchParams(history.location.search);
    urlParams.set('expand', chartKey);
    history.push(history.location.pathname + '?' + urlParams.toString());
  };
}

export { MetricsProps };
export default Metrics;
