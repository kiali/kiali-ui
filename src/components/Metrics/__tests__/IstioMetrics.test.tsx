import * as React from 'react';
import { mount, shallow, ReactWrapper } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router';

import IstioMetrics from '../IstioMetrics';
import * as API from '../../../services/Api';
import { MetricsObjectTypes, MonitoringDashboard, Chart } from '../../../types/Metrics';
import { store } from '../../../store/ConfigStore';

(window as any).SVGPathElement = a => a;
let mounted: ReactWrapper<any, any> | null;

const mockAPIToPromise = (func: keyof typeof API, obj: any): Promise<void> => {
  return new Promise((resolve, reject) => {
    jest.spyOn(API, func).mockImplementation(() => {
      return new Promise(r => {
        r({ data: obj });
        setTimeout(() => {
          try {
            resolve();
          } catch (e) {
            reject(e);
          }
        }, 1);
      });
    });
  });
};

const mockServiceDashboard = (dashboard: MonitoringDashboard): Promise<void> => {
  return mockAPIToPromise('getServiceDashboard', dashboard);
};

const mockWorkloadDashboard = (dashboard: MonitoringDashboard): Promise<void> => {
  return mockAPIToPromise('getWorkloadDashboard', dashboard);
};

const mockGrafanaInfo = (info: any): Promise<any> => {
  return mockAPIToPromise('getGrafanaInfo', info);
};

const createMetricChart = (name: string): Chart => {
  return {
    name: name,
    unit: 'B',
    spans: 12,
    metric: {
      matrix: [
        {
          metric: { __name__: name },
          values: [[1111, 5], [2222, 10]],
          name: ''
        }
      ]
    }
  };
};

const createHistogramChart = (name: string): Chart => {
  return {
    name: name,
    unit: 'B',
    spans: 12,
    histogram: {
      average: {
        matrix: [
          {
            metric: { __name__: name },
            values: [[1111, 10], [2222, 11]],
            name: name
          }
        ]
      },
      median: {
        matrix: [
          {
            metric: { __name__: name },
            values: [[1111, 20], [2222, 21]],
            name: name
          }
        ]
      },
      percentile95: {
        matrix: [
          {
            metric: { __name__: name },
            values: [[1111, 30], [2222, 31]],
            name: name
          }
        ]
      },
      percentile99: {
        matrix: [
          {
            metric: { __name__: name },
            values: [[1111, 40], [2222, 41]],
            name: name
          }
        ]
      }
    }
  };
};

describe('Metrics for a service', () => {
  beforeEach(() => {
    mounted = null;
  });
  afterEach(() => {
    if (mounted) {
      mounted.unmount();
    }
  });

  it('renders initial layout', () => {
    mockGrafanaInfo({});
    const wrapper = shallow(
      <Provider store={store}>
        <MemoryRouter>
          <Route
            render={props => (
              <IstioMetrics
                {...props}
                namespace="ns"
                object="svc"
                objectType={MetricsObjectTypes.SERVICE}
                direction={'inbound'}
                grafanaInfo={{
                  url: 'http://172.30.139.113:3000',
                  serviceDashboardPath: '/dashboard/db/istio-dashboard',
                  workloadDashboardPath: '/dashboard/db/istio-dashboard',
                  varService: 'var-service',
                  varNamespace: 'var-namespace',
                  varWorkload: 'var-workload'
                }}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('mounts and loads empty metrics', done => {
    const allMocksDone = [
      mockServiceDashboard({ title: 'foo', aggregations: [], charts: [] })
        .then(() => {
          mounted!.update();
          expect(mounted!.find('.card-pf')).toHaveLength(1);
          mounted!.find('.card-pf').forEach(pfCard => expect(pfCard.children().length === 0));
        })
        .catch(err => done.fail(err))
    ];
    Promise.all(allMocksDone).then(() => done());
    mounted = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Route
            render={props => (
              <IstioMetrics
                {...props}
                namespace="ns"
                object="svc"
                objectType={MetricsObjectTypes.SERVICE}
                direction={'inbound'}
                grafanaInfo={{
                  url: 'http://172.30.139.113:3000',
                  serviceDashboardPath: '/dashboard/db/istio-dashboard',
                  workloadDashboardPath: '/dashboard/db/istio-dashboard',
                  varService: 'var-service',
                  varNamespace: 'var-namespace',
                  varWorkload: 'var-workload'
                }}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );
  });

  it('mounts and loads full metrics', done => {
    const allMocksDone = [
      mockServiceDashboard({
        title: 'foo',
        aggregations: [],
        charts: [
          createMetricChart('m1'),
          createHistogramChart('m3'),
          createHistogramChart('m5'),
          createHistogramChart('m7')
        ]
      })
        .then(() => {
          mounted!.update();
          expect(mounted!.find('LineChart')).toHaveLength(4);
        })
        .catch(err => done.fail(err))
    ];
    Promise.all(allMocksDone).then(() => done());
    mounted = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Route
            render={props => (
              <IstioMetrics
                {...props}
                namespace="ns"
                object="svc"
                objectType={MetricsObjectTypes.SERVICE}
                direction={'inbound'}
                grafanaInfo={{
                  url: 'http://172.30.139.113:3000',
                  serviceDashboardPath: '/dashboard/db/istio-dashboard',
                  workloadDashboardPath: '/dashboard/db/istio-dashboard',
                  varService: 'var-service',
                  varNamespace: 'var-namespace',
                  varWorkload: 'var-workload'
                }}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );
  }, 10000); // Increase timeout for this test
});

describe('Inbound Metrics for a workload', () => {
  beforeEach(() => {
    mounted = null;
  });
  afterEach(() => {
    if (mounted) {
      mounted.unmount();
    }
  });

  it('renders initial layout', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <Route
          render={props => (
            <IstioMetrics
              {...props}
              namespace="ns"
              object="svc"
              objectType={MetricsObjectTypes.WORKLOAD}
              direction={'inbound'}
              grafanaInfo={{
                url: 'http://172.30.139.113:3000',
                serviceDashboardPath: '/dashboard/db/istio-dashboard',
                workloadDashboardPath: '/dashboard/db/istio-dashboard',
                varService: 'var-service',
                varNamespace: 'var-namespace',
                varWorkload: 'var-workload'
              }}
            />
          )}
        />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('mounts and loads empty metrics', done => {
    const allMocksDone = [
      mockWorkloadDashboard({ title: 'foo', aggregations: [], charts: [] })
        .then(() => {
          mounted!.update();
          expect(mounted!.find('.card-pf')).toHaveLength(1);
          mounted!.find('.card-pf').forEach(pfCard => expect(pfCard.children().length === 0));
        })
        .catch(err => done.fail(err))
    ];
    Promise.all(allMocksDone).then(() => done());
    mounted = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Route
            render={props => (
              <IstioMetrics
                {...props}
                namespace="ns"
                object="svc"
                objectType={MetricsObjectTypes.WORKLOAD}
                direction={'inbound'}
                grafanaInfo={{
                  url: 'http://172.30.139.113:3000',
                  serviceDashboardPath: '/dashboard/db/istio-dashboard',
                  workloadDashboardPath: '/dashboard/db/istio-dashboard',
                  varService: 'var-service',
                  varNamespace: 'var-namespace',
                  varWorkload: 'var-workload'
                }}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );
  });

  it('mounts and loads full metrics', done => {
    const allMocksDone = [
      mockWorkloadDashboard({
        title: 'foo',
        aggregations: [],
        charts: [
          createMetricChart('m1'),
          createHistogramChart('m3'),
          createHistogramChart('m5'),
          createHistogramChart('m7')
        ]
      })
        .then(() => {
          mounted!.update();
          expect(mounted!.find('LineChart')).toHaveLength(4);
        })
        .catch(err => done.fail(err))
    ];
    Promise.all(allMocksDone).then(() => done());
    mounted = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Route
            render={props => (
              <IstioMetrics
                {...props}
                namespace="ns"
                object="svc"
                objectType={MetricsObjectTypes.WORKLOAD}
                direction={'inbound'}
                grafanaInfo={{
                  url: 'http://172.30.139.113:3000',
                  serviceDashboardPath: '/dashboard/db/istio-dashboard',
                  workloadDashboardPath: '/dashboard/db/istio-dashboard',
                  varService: 'var-service',
                  varNamespace: 'var-namespace',
                  varWorkload: 'var-workload'
                }}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );
  }, 10000); // Increase timeout for this test
});
