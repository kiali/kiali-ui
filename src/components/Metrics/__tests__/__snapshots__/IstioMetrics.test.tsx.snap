// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inbound Metrics for a workload renders initial layout 1`] = `
<Route
  render={[Function]}
/>
`;

exports[`Metrics for a service renders initial layout 1`] = `
<MemoryRouter>
  <Route
    render={[Function]}
  />
</MemoryRouter>
`;
